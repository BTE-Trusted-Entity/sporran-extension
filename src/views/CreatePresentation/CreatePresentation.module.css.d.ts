export const __esModule: true;
export const container: string;
export const heading: string;
export const subline: string;
export const credentialsList: string;
export const buttonsLine: string;
export const cancel: string;
export const download: string;
