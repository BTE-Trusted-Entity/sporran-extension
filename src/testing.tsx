export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
