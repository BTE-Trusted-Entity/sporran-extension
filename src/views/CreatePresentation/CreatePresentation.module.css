.container {
  composes: container from '../../components/View/View.module.css';
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
}

.subline {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0.25rem 0 0.5rem;
}

.credentialContainer {
  composes: fullWidthContainer from '../../components/View/View.module.css';
}

.buttonsLine {
  composes: buttonsLine from '../../components/Button/Button.module.css';
}

.cancel {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
}

.download {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}
