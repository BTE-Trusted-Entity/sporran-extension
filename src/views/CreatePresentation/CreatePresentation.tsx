import { useState } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import { browser } from 'webextension-polyfill-ts';
import { find } from 'lodash-es';

import * as styles from './CreatePresentation.module.css';

import { Identity } from '../../utilities/identities/types';
import { getIdentityDid } from '../../utilities/identities/identities';
import {
  getUnsignedPresentationDownload,
  useIdentityCredentials,
} from '../../utilities/credentials/credentials';
import { LinkBack } from '../../components/LinkBack/LinkBack';
import { Stats } from '../../components/Stats/Stats';
import { PresentCredentialCard } from '../../components/CredentialCard/PresentCredentialCard';

interface Props {
  identity: Identity;
}

export function CreatePresentation({ identity }: Props): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const { goBack } = useHistory();
  const params = useParams() as { hash: string };
  const rootHash = params.hash;

  const did = getIdentityDid(identity);
  const sporranCredentials = useIdentityCredentials(did);
  const sporranCredential = find(sporranCredentials, {
    credential: { rootHash },
  });
  const [checked, setChecked] = useState<string[]>([]);

  if (!sporranCredentials) {
    return null; // storage data pending
  }

  if (!sporranCredential) {
    throw new Error(`Credential not found: ${rootHash}`);
  }

  const { name, url } = getUnsignedPresentationDownload(
    sporranCredential,
    checked,
  );

  return (
    <section className={styles.container}>
      <h1 className={styles.heading}>{t('view_CreatePresentation_title')}</h1>
      <p className={styles.subline}>{t('view_CreatePresentation_subline')}</p>

      <div className={styles.credentialContainer}>
        <PresentCredentialCard
          sporranCredential={sporranCredential}
          onSelect={setChecked}
        />
      </div>

      <p className={styles.buttonsLine}>
        <button onClick={goBack} className={styles.cancel}>
          {t('common_action_cancel')}
        </button>
        <a download={name} href={url} className={styles.download}>
          {t('view_CreatePresentation_download')}
        </a>
      </p>

      <LinkBack />
      <Stats />
    </section>
  );
}
