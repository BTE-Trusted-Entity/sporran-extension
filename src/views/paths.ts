export { generatePath } from 'react-router-dom';

export const paths = {
  home: '/',
  settings: '/settings',
  access: '/access',
  identity: {
    base: '/identity/',
    add: '/identity/add',
    create: {
      start: '/identity/create',
      backup: '/identity/create/backup',
      verify: '/identity/create/verify',
      password: '/identity/create/password',
    },
    import: {
      start: '/identity/import',
      password: '/identity/import/password',
    },

    overview: '/identity/:address/:type(created|imported|pwreset)?',
    reset: {
      start: '/identity/:address/reset',
      password: '/identity/:address/reset/password',
    },
    send: {
      start: '/identity/:address/send',
      review: '/identity/:address/send/review',
      warning: '/identity/:address/send/warning',
    },
    receive: '/identity/:address/receive',
    credentials: {
      base: '/identity/:address/credentials',
      presentation: '/identity/:address/credentials/:hash/presentation',
    },
    remove: '/identity/:address/remove',
    did: {
      upgrade: {
        start: '/identity/:address/did/upgrade',
        kilt: '/identity/:address/did/upgrade/kilt',
        euro: '/identity/:address/did/upgrade/euro',
      },
      manage: {
        start: '/identity/:address/did',
        downgrade: {
          base: '/identity/:address/did/downgrade',
          warning: {
            web3name: '/identity/:address/did/downgrade/warning/web3name',
            credentials: '/identity/:address/did/downgrade/warning/credentials',
          },
          sign: '/identity/:address/did/downgrade/sign',
        },
        endpoints: {
          start: '/identity/:address/did/endpoints',
          edit: '/identity/:address/did/endpoints/:id?',
          sign: '/identity/:address/did/endpoints/sign',
        },
      },
      repair: '/identity/:address/did/repair',
    },
    web3name: {
      create: {
        base: '/identity/:address/web3name/create',
        info: '/identity/:address/web3name/create/info',
        form: '/identity/:address/web3name/create/form',
        choose: '/identity/:address/web3name/create/choose/:web3name',
        kilt: '/identity/:address/web3name/create/kilt/:web3name',
        euro: '/identity/:address/web3name/create/euro/:web3name',
      },
      manage: {
        start: '/identity/:address/web3name',
        remove: '/identity/:address/web3name/remove',
      },
    },
  },
  popup: {
    base: '/popup',
    import: '/popup/import',
    access: '/popup/access',
    claim: '/identity/:address/claim',
    save: '/popup/save',
    share: {
      start: '/popup/share',
      sign: '/popup/share/sign',
    },
    sign: '/popup/sign',
    signRaw: '/popup/signRaw',
    signDid: {
      start: '/identity/:address/did/sign',
      sign: '/identity/:address/did/sign/password',
      credentials: '/identity/:address/did/sign/credentials',
    },
    signDidExtrinsic: '/identity/:address/did/signExtrinsic',
    createDid: '/identity/:address/did/create',
  },
};
