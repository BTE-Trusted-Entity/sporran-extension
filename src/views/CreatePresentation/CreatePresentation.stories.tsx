import { Meta } from '@storybook/react';
import { MemoryRouter, Route } from 'react-router-dom';

import { credentialsMock } from '../../utilities/credentials/CredentialsProvider.mock';
import { moreIdentitiesMock as identities } from '../../utilities/identities/IdentitiesProvider.mock';
import { paths } from '../paths';

import { CreatePresentation } from './CreatePresentation';

export default {
  title: 'Views/CreatePresentation',
  component: CreatePresentation,
} as Meta;

const identity = identities['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo'];

export function Template(): JSX.Element {
  return (
    <MemoryRouter
      initialEntries={[
        `/identity/${identity.address}/credentials/${credentialsMock[12].credential.rootHash}/presentation`,
      ]}
    >
      <Route path={paths.identity.credentials.presentation}>
        <CreatePresentation identity={identity} />
      </Route>
    </MemoryRouter>
  );
}
