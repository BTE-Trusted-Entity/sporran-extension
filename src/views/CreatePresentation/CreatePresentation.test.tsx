import { MemoryRouter, Route } from 'react-router-dom';

import { identitiesMock, render } from '../../testing/testing';
import { credentialsMock } from '../../utilities/credentials/CredentialsProvider.mock';
import {
  getUnsignedPresentationDownload,
  useIdentityCredentials,
} from '../../utilities/credentials/credentials';
import { paths } from '../paths';

import { CreatePresentation } from './CreatePresentation';

jest.mock('../../utilities/credentials/credentials');
jest.mocked(useIdentityCredentials).mockReturnValue(credentialsMock);
jest
  .mocked(getUnsignedPresentationDownload)
  .mockReturnValue({ name: 'name', url: 'data:' });

const identity =
  identitiesMock['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1'];

describe('CreatePresentation', () => {
  it('should match the snapshot', async () => {
    const { container } = render(
      <MemoryRouter
        initialEntries={[
          `/identity/${identity.address}/credentials/${credentialsMock[0].credential.rootHash}/presentation`,
        ]}
      >
        <Route path={paths.identity.credentials.presentation}>
          <CreatePresentation identity={identity} />
        </Route>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });
});
