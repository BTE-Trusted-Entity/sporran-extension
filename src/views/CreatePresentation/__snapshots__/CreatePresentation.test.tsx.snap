// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePresentation should match the snapshot 1`] = `
<div>
  <section
    class="container"
  >
    <h1
      class="heading"
    >
      Customize Credential
    </h1>
    <p
      class="subline"
    >
      Select the information to disclose
    </p>
    <div
      class="credentialContainer"
    >
      <form
        class="credential"
      >
        <section
          class="expanded"
        >
          <section>
            <dl
              class="details"
            >
              <div
                class="detail"
              >
                <dt
                  class="detailName"
                >
                  Credential name
                </dt>
                <dd
                  class="detailValue"
                >
                  Email Credential
                </dd>
              </div>
              <div
                class="detail"
              >
                <dt
                  class="detailName"
                >
                  Status
                </dt>
                <dd
                  class="detailValue"
                >
                  Attested
                </dd>
              </div>
              <div
                class="detail"
              >
                <dt
                  class="detailName"
                >
                  Email
                </dt>
                <dd
                  class="detailValue"
                >
                  <label
                    class="shareLabel"
                  >
                    <input
                      checked=""
                      class="share"
                      name="Email"
                      type="checkbox"
                    />
                    <span />
                    mockEmail@mock.mock
                  </label>
                </dd>
              </div>
            </dl>
          </section>
          <h4
            class="technical"
          >
            Technical details
          </h4>
          <dl
            class="details"
          >
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                CType
              </dt>
              <dd
                class="detailValue"
              >
                Email
              </dd>
            </div>
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                Attester
              </dt>
              <dd
                class="detailValue"
              >
                Trusted Entity attester
              </dd>
            </div>
            <div
              class="hash"
            >
              <dt
                class="detailName"
              >
                Credential hash
              </dt>
              <dd
                class="detailValue"
              >
                0xbe7fcc7aa0186b05fd116c100e2d673fb951163693788640ce6032ad2f700dae
              </dd>
            </div>
          </dl>
        </section>
      </form>
    </div>
    <p
      class="buttonsLine"
    >
      <button
        class="cancel"
      >
        Cancel
      </button>
      <a
        class="download"
        download="name"
        href="data:"
      >
        Download
      </a>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </section>
</div>
`;
